### ISSUE TITLE
`textDocument/didChange`  Did not follow the LSP protocol

### ISSUE THREAD

**Issue by vincentLiuxiang** [TypeFox/monaco-languageclient#45]
hi all, I just want to try lsp in monaco, and I was very excited to find this project.
However, When I tried to analyze the websocket data package, I found a bug.

![image](https://user-images.githubusercontent.com/10772305/35496302-704d6a0c-0500-11e8-8187-9928e8344613.png)

then, I tried to analyze theia-ide, It works fine.

![image](https://user-images.githubusercontent.com/10772305/35496345-b8ae9640-0500-11e8-938a-d8d15867d77f.png)

what should I do to solve this problem?

**Comment by akosyakov** [comment]
Hi @vincentLiuxiang, could you be more specific that does not follow the LSP?

**Comment by vincentLiuxiang** [comment]
@akosyakov 
You can look at the red square inside the pictures.
[textDocument/didChange](https://github.com/Microsoft/language-server-protocol/blob/gh-pages/specification.md#didchangetextdocument-notification-arrow_right)

In LSP, when the method is `textDocument/didChange`, the params must be DidChangeTextDocumentParams
```
interface DidChangeTextDocumentParams {
	/**
	 * The document that did change. The version number points
	 * to the version after all provided content changes have
	 * been applied.
	 */
	textDocument: VersionedTextDocumentIdentifier;

	/**
	 * The actual content changes. The content changes describe single state changes
	 * to the document. So if there are two content changes c1 and c2 for a document 
	 * in state S10 then c1 move the document to S11 and c2 to S12.
	 */
	contentChanges: TextDocumentContentChangeEvent[];
}

/**
 * An event describing a change to a text document. If range and rangeLength are omitted
 * the new text is considered to be the full content of the document.
 */
interface TextDocumentContentChangeEvent {
	/**
	 * The range of the document that changed.
	 */
	range?: Range;

	/**
	 * The length of the range that got replaced.
	 */
	rangeLength?: number;

	/**
	 * The new text of the range/document.
	 */
	text: string;
}
```

but `monaco-languageclient` send data to sever like this:
```
{
    "method": "textDocument/didChange",
    "params": {
         "contentChanges": [
                 {  "text": "{↵    "$schema": "http://json.schemastore.org/coffeelint",↵    "line_endings": "unix"↵}aaa"  }
         ]
     }
}
```
the difference is in `contentChanges`.

**Comment by akosyakov** [comment]
It depends on a language server, the json server used in the example is not able to handle incremental changes, so the client always sends the whole document. A server can communicate it as a part of exchanging initialization capabilities. Look for `TextDocumentSyncKind` in the specification.

**Comment by vincentLiuxiang** [comment]
### monaco-languageclient

![15](https://user-images.githubusercontent.com/10772305/35500121-53334012-0510-11e8-8af9-f6e90a771a28.gif)

### theia-ide
![15](https://user-images.githubusercontent.com/10772305/35500274-ef217e1c-0510-11e8-8af8-1a5c59ec8aa2.gif)

**Comment by vincentLiuxiang** [comment]
so, if My server support handle  incremental changes, I can get the same result like the theia-ide?

what's more could you give an example that is closer to the production environment? java, js and so on?

**Comment by vincentLiuxiang** [comment]
@akosyakov I am very new in LSP, I have Look for TextDocumentSyncKind in the specification. Thank you very much!

**Comment by akosyakov** [comment]
@vincentLiuxiang You are welcome, I am closing it.

**Comment by Louis-7** [comment]
I know this issue has been closed but I saw the screen shot, there is a "Content-Length: 25x" at the beginning of the data return by web-socket which as same as my project which use the language server provide by monaco-languageclient. This "Content-Length" will cause an error at language server side. But I didn't see it in the example given by monaco-languageclient. I'm wondering why I have this "Content-Length" and will it crash the hole process?
